### ISSUE TITLE
Problem in getDataFromTree, React Component detection

### ISSUE THREAD

**Issue by edorivai** [apollographql/react-apollo#1595]
**Intended outcome:**

`getDataFromTree` should not crash when passed a valid React tree.

**Actual outcome:**

`getDataFromTree` crashes when the `<Link>` component from `react-scroll` is rendered.
Error: `error: Cannot call a class as a function TypeError: Cannot call a class as a function`

@leoasis was quick to respond in the Slack channel, and we narrowed it down to the following commit:
https://github.com/apollographql/react-apollo/commit/1dff95968d1dad99ee91665c7c0a5c35274e0d22
in combination with the following line from `react-scroll`:
https://github.com/fisshy/react-scroll/blob/master/modules/components/Link.js#L7

The commit checks for `Component.prototype.render`, however, in `react-scroll` the render function is defined on the instance, instead of the prototype.

Something along the lines of this works in the case of `react-scroll`:
```js
if (Component.prototype && (Component.prototype.render || Component.prototype.isReactComponent)) {
```

But I have no experience with preact, so not sure if this covers all of the possible cases.

**Version**

* react-apollo@2.1.0-beta.0

**Comment by jbaxleyiii** [comment]
crazy! Seems like a good change. Would you be open to a PR for it?

**Comment by developit** [comment]
The suggestion would cover all cases.

**Comment by edorivai** [comment]
Should be fixed in #1604
